### PR TITLE
Add --useReadOnlyType for TS

### PR DESCRIPTION
**Pull request by mike-marcacci**
- [x] Update CHANGELOG.md\* with your change (include reference to issue & this PR)
- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass

Please note that the checks currently fail due to a flaw in the type definitions for `babel-types`, which is fixed in [this PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34926). Once that lands, the tests will pass.

### CODE REVIEW

**Comment by mike-marcacci** [comment]
OK – the upstream PR has been merged, and all tests pass.

**Comment by mike-marcacci** [comment]
Bump.

**Comment by mike-marcacci** [comment]
Hey, can somebody look at this?

**Comment by mike-marcacci** [comment]
Just updated to fix merge conflicts.

**Comment by trevor-scheer** [comment]
Fun fact, if you update and see some funny business with the tests (azure 8 & 10 failing unexpectedly), feel free to ignore. Tests are a bit flaky for now, and a rerun usually resolves the problem.

File: packages/apollo-codegen-typescript/src/language.ts [hunk 1]
```diff
@@ -17,7 +17,7 @@ export type ObjectProperty = {
 };
 
 export interface TypescriptCompilerOptions extends CompilerOptions {
-  // Leaving this open for Typescript only compiler options
+  useReadOnlyTypes: boolean;
```
[the remark this answers is not in the document]
Comment (mike-marcacci) [reply packages/apollo-codegen-typescript/src/language.ts]: A yes, just preserved this for symmetry with the flow codegen. Will remove.

File: packages/apollo/src/commands/client/codegen.ts [hunk 2]
```diff
@@ -77,11 +77,12 @@ export default class Generate extends ClientCommand {
     useFlowExactObjects: flags.boolean({
       description: "Use Flow exact objects for generated types [flow only]"
     }),
-    useFlowReadOnlyTypes: flags.boolean({
```
[the remark this answers is not in the document]
Comment (mike-marcacci) [reply packages/apollo/src/commands/client/codegen.ts]: I figured this was the most likely request :) Will add shortly.

**Comment by mike-marcacci** [comment]
I've made the requested changes, but can no longer run `npm i` locally on either master or this rebased branch – the lerna hook fails multiple times with:

```
> lerna run build --stream --scope apollo-env && npm run compile && lerna run build --stream --ignore apollo-env

lerna notice cli v3.13.4
lerna info versioning independent
lerna info filter [ 'apollo-env' ]
lerna info Executing command in 1 package: "npm run build"
apollo-env: > apollo-env@0.5.0 prebuild /Users/mike/Code/apollo-tooling/packages/apollo-env
apollo-env: > npm run clean
apollo-env: > apollo-env@0.5.0 clean /Users/mike/Code/apollo-tooling/packages/apollo-env
apollo-env: > npx rimraf lib
apollo-env: > apollo-env@0.5.0 build /Users/mike/Code/apollo-tooling/packages/apollo-env
apollo-env: > tsc --project ./src/fetch && node clone-types.js && cd src && tsc
apollo-env: copied: src/fetch/fetch.d.ts -> lib/fetch/fetch.d.ts
apollo-env: copied: src/fetch/global.d.ts -> lib/fetch/global.d.ts
apollo-env: copied: src/fetch/index.d.ts -> lib/fetch/index.d.ts
apollo-env: copied: src/fetch/url.d.ts -> lib/fetch/url.d.ts
lerna success run Ran npm script 'build' in 1 package in 2.7s:
lerna success - apollo-env

> apollo-tooling-monorepo@ compile /Users/mike/Code/apollo-tooling
> tsc --build tsconfig.json

error TS2688: Cannot find type definition file for 'apollo-env'.
```

...so I can't build or run tests locally.

I manually created the snapshot so that the new test passes when run on CircleCI so this PR should be good to go.

(But you'll still probably want to try cloning a fresh copy of the repo to see if there's a more general problem.)

**Comment by trevor-scheer** [comment]
Looks great! An `npm run clean && npm i` might resolve the build issue you're seeing in development. I added a commit to update the snapshot (looks like it was slightly off), but otherwise this is awesome. I'll merge and release right now 😄

Thanks again for the work on this, including the upstream types fix!

**Comment by mike-marcacci** [comment]
Awesome - Thanks so much for the great review and snappy followup!

**Comment by trevor-scheer** [comment]
Released in `apollo@2.12.0` 🎉

**Pull request merged**